File: ios/Base/RtcChannel.swift
//
//  ArRtcChannel.swift
//  ar_rtc_engine
//
//  Created by 余生丶 on 2020/10/29.
//

import Foundation
import ARtcKit

protocol RtcChannelInterface:
        RtcChannelAudioInterface,
        RtcChannelVideoInterface,
        RtcChannelDualStreamInterface {
    func create(_ params: NSDictionary, _ callback: Callback)

    func destroy(_ params: NSDictionary, _ callback: Callback)

    func setClientRole(_ params: NSDictionary, _ callback: Callback)

    func joinChannel(_ params: NSDictionary, _ callback: Callback)

    func leaveChannel(_ params: NSDictionary, _ callback: Callback)

    func renewToken(_ params: NSDictionary, _ callback: Callback)

    func publish(_ params: NSDictionary, _ callback: Callback)

    func unpublish(_ params: NSDictionary, _ callback: Callback)
}

protocol RtcChannelAudioInterface {
    func adjustUserPlaybackSignalVolume(_ params: NSDictionary, _ callback: Callback)

    func muteRemoteAudioStream(_ params: NSDictionary, _ callback: Callback)

    func muteAllRemoteAudioStreams(_ params: NSDictionary, _ callback: Callback)

    func setDefaultMuteAllRemoteAudioStreams(_ params: NSDictionary, _ callback: Callback)
}

protocol RtcChannelVideoInterface {
    func muteRemoteVideoStream(_ params: NSDictionary, _ callback: Callback)

    func muteAllRemoteVideoStreams(_ params: NSDictionary, _ callback: Callback)

    func setDefaultMuteAllRemoteVideoStreams(_ params: NSDictionary, _ callback: Callback)
}

protocol RtcChannelDualStreamInterface {
    func setRemoteVideoStreamType(_ params: NSDictionary, _ callback: Callback)

    func setRemoteDefaultVideoStreamType(_ params: NSDictionary, _ callback: Callback)
}


@objc
class RtcChannelManager: NSObject, RtcChannelInterface {
    private var emitter: (_ methodName: String, _ data: Dictionary<String, Any?>?) -> Void
    private var rtcChannelMap = [String: ARtcChannel]()
    private var rtcChannelDelegateMap = [String: RtcChannelEventHandler]()

    init(_ emitter: @escaping (_ methodName: String, _ data: Dictionary<String, Any?>?) -> Void) {
        self.emitter = emitter
    }

    func Release() {
        rtcChannelMap.forEach { key, value in
            value.destroy()
        }
        rtcChannelMap.removeAll()
        rtcChannelDelegateMap.removeAll()
    }

    subscript(channelId: String) -> ARtcChannel? {
        get {
            return rtcChannelMap[channelId]
        }
    }

    @objc func create(_ params: NSDictionary, _ callback: Callback) {
        callback.resolve(params["engine"] as? ARtcEngineKit) { [weak self] it in
            if let rtcChannel = it.createRtcChannel(params["channelId"] as! String) {
                let delegate = RtcChannelEventHandler() { [weak self] methodName, data in
                    self?.emitter(methodName, data)
                }
                rtcChannel.setRtcChannelDelegate(delegate)
                self?.rtcChannelMap[rtcChannel.getId()!] = rtcChannel
                self?.rtcChannelDelegateMap[rtcChannel.getId()!] = delegate
            }
            return nil
        }
    }

    @objc func destroy(_ params: NSDictionary, _ callback: Callback) {
        var code: Int32? = -Int32(ARErrorCode.notInitialized.rawValue)
        if let it = self[params["channelId"] as! String] {
            code = rtcChannelMap.removeValue(forKey: it.getId()!)?.destroy()
        }
        callback.code(code)
    }

    @objc func setClientRole(_ params: NSDictionary, _ callback: Callback) {
        callback.code(self[params["channelId"] as! String]?.setClientRole(ARClientRole(rawValue: params["role"] as! Int)!))
    }

    @objc func joinChannel(_ params: NSDictionary, _ callback: Callback) {
        callback.code(self[params["channelId"] as! String]?.join(byToken: params["token"] as? String,uid: params["optionalUid"] as? String, options: mapToChannelMediaOptions(params["options"] as! Dictionary)))
    }

    @objc func leaveChannel(_ params: NSDictionary, _ callback: Callback) {
        callback.code(self[params["channelId"] as! String]?.leave())
    }

    @objc func renewToken(_ params: NSDictionary, _ callback: Callback) {
        callback.code(self[params["channelId"] as! String]?.renewToken(params["token"] as! String))
    }

    @objc func publish(_ params: NSDictionary, _ callback: Callback) {
        callback.code(self[params["channelId"] as! String]?.publish())
    }

    @objc func unpublish(_ params: NSDictionary, _ callback: Callback) {
        callback.code(self[params["channelId"] as! String]?.unpublish())
    }

    @objc func adjustUserPlaybackSignalVolume(_ params: NSDictionary, _ callback: Callback) {
        callback.code(self[params["channelId"] as! String]?.adjustUserPlaybackSignalVolume(params["uid"] as! String, volume: params["volume"] as! Int32))
    }

    @objc func muteRemoteAudioStream(_ params: NSDictionary, _ callback: Callback) {
        callback.code(self[params["channelId"] as! String]?.muteRemoteAudioStream(params["uid"] as! String, mute: params["muted"] as! Bool))
    }

    @objc func muteAllRemoteAudioStreams(_ params: NSDictionary, _ callback: Callback) {
        callback.code(self[params["channelId"] as! String]?.muteAllRemoteAudioStreams(params["muted"] as! Bool))
    }

    @objc func setDefaultMuteAllRemoteAudioStreams(_ params: NSDictionary, _ callback: Callback) {
        callback.code(self[params["channelId"] as! String]?.setDefaultMuteAllRemoteAudioStreams(params["muted"] as! Bool))
    }

    @objc func muteRemoteVideoStream(_ params: NSDictionary, _ callback: Callback) {
        callback.code(self[params["channelId"] as! String]?.muteRemoteVideoStream(params["uid"] as! String, mute: params["muted"] as! Bool))
    }

    @objc func muteAllRemoteVideoStreams(_ params: NSDictionary, _ callback: Callback) {
        callback.code(self[params["channelId"] as! String]?.muteAllRemoteVideoStreams(params["muted"] as! Bool))
    }

    @objc func setDefaultMuteAllRemoteVideoStreams(_ params: NSDictionary, _ callback: Callback) {
        callback.code(self[params["channelId"] as! String]?.setDefaultMuteAllRemoteVideoStreams(params["muted"] as! Bool))
    }

    @objc func setRemoteVideoStreamType(_ params: NSDictionary, _ callback: Callback) {
        callback.code(self[params["channelId"] as! String]?.setRemoteVideoStream(params["uid"] as! String, type: ARVideoStreamType(rawValue: params["streamType"] as! Int)!))
    }

    @objc func setRemoteDefaultVideoStreamType(_ params: NSDictionary, _ callback: Callback) {
        callback.code(self[params["channelId"] as! String]?.setRemoteDefaultVideoStreamType(ARVideoStreamType(rawValue: params["streamType"] as! Int)!))
    }
}

File: android/src/main/java/org/ar/rtc/base/RtcChannel.kt
package org.ar.rtc.base

import org.ar.rtc.Constants
import org.ar.rtc.RtcChannel
import org.ar.rtc.RtcEngine
import java.util.*

class IRtcChannel {
    interface RtcChannelInterface : RtcAudioInterface, RtcVideoInterface, RtcVoicePositionInterface,
            RtcPublishStreamInterface, RtcMediaRelayInterface, RtcDualStreamInterface,
            RtcFallbackInterface, RtcMediaMetadataInterface, RtcEncryptionInterface,
            RtcInjectStreamInterface, RtcStreamMessageInterface {
        fun create(params: Map<String, *>, callback: Callback)

        fun destroy(params: Map<String, *>, callback: Callback)

        fun setClientRole(params: Map<String, *>, callback: Callback)

        fun joinChannel(params: Map<String, *>, callback: Callback)

        fun leaveChannel(params: Map<String, *>, callback: Callback)

        fun renewToken(params: Map<String, *>, callback: Callback)

        fun getConnectionState(params: Map<String, *>, callback: Callback)

        fun publish(params: Map<String, *>, callback: Callback)

        fun unpublish(params: Map<String, *>, callback: Callback)

        fun getCallId(params: Map<String, *>, callback: Callback)
    }

    interface RtcAudioInterface {
        fun adjustUserPlaybackSignalVolume(params: Map<String, *>, callback: Callback)

        fun muteRemoteAudioStream(params: Map<String, *>, callback: Callback)

        fun muteAllRemoteAudioStreams(params: Map<String, *>, callback: Callback)

        fun setDefaultMuteAllRemoteAudioStreams(params: Map<String, *>, callback: Callback)
    }

    interface RtcVideoInterface {
        fun muteRemoteVideoStream(params: Map<String, *>, callback: Callback)

        fun muteAllRemoteVideoStreams(params: Map<String, *>, callback: Callback)

        fun setDefaultMuteAllRemoteVideoStreams(params: Map<String, *>, callback: Callback)
    }

    interface RtcVoicePositionInterface {
    }

    interface RtcPublishStreamInterface {
    }

    interface RtcMediaRelayInterface {
    }

    interface RtcDualStreamInterface {
        fun setRemoteVideoStreamType(params: Map<String, *>, callback: Callback)

        fun setRemoteDefaultVideoStreamType(params: Map<String, *>, callback: Callback)
    }

    interface RtcFallbackInterface {
    }

    interface RtcMediaMetadataInterface {
    }

    interface RtcEncryptionInterface {
    }

    interface RtcInjectStreamInterface {
    }

    interface RtcStreamMessageInterface {
    }
}

class RtcChannelManager(
        private val emit: (methodName: String, data: Map<String, Any?>?) -> Unit
) : IRtcChannel.RtcChannelInterface {
    private val rtcChannelMap = Collections.synchronizedMap(mutableMapOf<String, RtcChannel>())

    fun release() {
        rtcChannelMap.forEach { it.value.destroy() }
        rtcChannelMap.clear()
    }

    operator fun get(channelId: String): RtcChannel? {
        return rtcChannelMap[channelId]
    }

    override fun create(params: Map<String, *>, callback: Callback) {
        callback.resolve(params["engine"] as RtcEngine) { e ->
            e.createRtcChannel(params["channelId"] as String)?.let {
                it.setRtcChannelEventHandler(RtcChannelEventHandler { methodName, data -> emit(methodName, data) })
                rtcChannelMap[it.channelId()] = it
            }
            Unit
        }
    }

    override fun destroy(params: Map<String, *>, callback: Callback) {
        var code: Int? = -Constants.ERR_NOT_INITIALIZED
        this[params["channelId"] as String]?.let {
            code = rtcChannelMap.remove(it.channelId())?.destroy()
        }
        callback.code(code)
    }

    override fun setClientRole(params: Map<String, *>, callback: Callback) {
        callback.code(this[params["channelId"] as String]?.setClientRole((params["role"] as Number).toInt()))
    }

    override fun joinChannel(params: Map<String, *>, callback: Callback) {
        callback.code(this[params["channelId"] as String]?.joinChannel(params["token"] as? String, params["optionalInfo"] as? String, (params["optionalUid"] as String).toString(), mapToChannelMediaOptions(params["options"] as Map<*, *>)))
    }


    override fun leaveChannel(params: Map<String, *>, callback: Callback) {
        callback.code(this[params["channelId"] as String]?.leaveChannel())
    }

    override fun renewToken(params: Map<String, *>, callback: Callback) {
        callback.code(this[params["channelId"] as String]?.renewToken(params["token"] as String))
    }

    override fun getConnectionState(params: Map<String, *>, callback: Callback) {
        callback.resolve(this[params["channelId"] as String]) { it.connectionState }
    }

    override fun publish(params: Map<String, *>, callback: Callback) {
        callback.code(this[params["channelId"] as String]?.publish())
    }

    override fun unpublish(params: Map<String, *>, callback: Callback) {
        callback.code(this[params["channelId"] as String]?.unpublish())
    }

    override fun getCallId(params: Map<String, *>, callback: Callback) {
        callback.resolve(this[params["channelId"] as String]) { it.callId }
    }

    override fun adjustUserPlaybackSignalVolume(params: Map<String, *>, callback: Callback) {
        callback.code(this[params["channelId"] as String]?.adjustUserPlaybackSignalVolume((params["uid"] as String).toString(), (params["volume"] as Number).toInt()))
    }

    override fun muteRemoteAudioStream(params: Map<String, *>, callback: Callback) {
        callback.code(this[params["channelId"] as String]?.muteRemoteAudioStream((params["uid"] as String).toString(), params["muted"] as Boolean))
    }

    override fun muteAllRemoteAudioStreams(params: Map<String, *>, callback: Callback) {
        callback.code(this[params["channelId"] as String]?.muteAllRemoteAudioStreams(params["muted"] as Boolean))
    }

    override fun setDefaultMuteAllRemoteAudioStreams(params: Map<String, *>, callback: Callback) {
        callback.code(this[params["channelId"] as String]?.setDefaultMuteAllRemoteAudioStreams(params["muted"] as Boolean))
    }

    override fun muteRemoteVideoStream(params: Map<String, *>, callback: Callback) {
        callback.code(this[params["channelId"] as String]?.muteRemoteVideoStream((params["uid"] as String).toString(), params["muted"] as Boolean))
    }

    override fun muteAllRemoteVideoStreams(params: Map<String, *>, callback: Callback) {
        callback.code(this[params["channelId"] as String]?.muteAllRemoteVideoStreams(params["muted"] as Boolean))
    }

    override fun setDefaultMuteAllRemoteVideoStreams(params: Map<String, *>, callback: Callback) {
        callback.code(this[params["channelId"] as String]?.setDefaultMuteAllRemoteVideoStreams(params["muted"] as Boolean))
    }



    override fun setRemoteVideoStreamType(params: Map<String, *>, callback: Callback) {
        callback.code(this[params["channelId"] as String]?.setRemoteVideoStreamType((params["uid"] as String).toString(), (params["streamType"] as Number).toInt()))
    }

    override fun setRemoteDefaultVideoStreamType(params: Map<String, *>, callback: Callback) {
        callback.code(this[params["channelId"] as String]?.setRemoteDefaultVideoStreamType((params["streamType"] as Number).toInt()))
    }








}
